src/Interpreters/Access/InterpreterCreateRoleQuery.cpp

```diff
@@ -78,12 +93,27 @@ BlockIO InterpreterCreateRoleQuery::execute()
             new_roles.emplace_back(std::move(new_role));
         }
 
+        if (!query.storage_name.empty())
+        {
+            for (const auto & name : query.names)
+            {
+                auto id = access_control.find<Role>(name);
+
+                if (!id)
+                    continue;
+
+                auto another_storage_ptr = access_control.findStorage(*id);
+                if (another_storage_ptr != storage_ptr)
+                    throw Exception(ErrorCodes::ACCESS_ENTITY_ALREADY_EXISTS, "Role {} already exists in storage {}", name, another_storage_ptr->getStorageName());
+            }
+        }
+
         if (query.if_not_exists)
-            access_control.tryInsert(new_roles);
+            storage->tryInsert(new_roles);
         else if (query.or_replace)
-            access_control.insertOrReplace(new_roles);
+            storage->insertOrReplace(new_roles);
         else
-            access_control.insert(new_roles);
+            storage->insert(new_roles);
```

Comment:
It looks like it would be useful also to have a command to move a user/role:

```
MOVE USER <user_name> TO <storage_name>
```

Of course it won't work for users.xml because it's readonly, but it could be useful for other storages.